### ISSUE TITLE
Replace Firebase/Firstore with something better

### ISSUE THREAD

**Issue by bvaughn** [bvaughn/personal-logger#27]
In general, I've found Firebase to be awkward to work with- from authentication to querying.

Its support for compound queries is very limited. For example, it doesn't support logical OR or the "!=" operator, pushing the burden onto the client.

Its authentication model feels awkward too: Sessions aren't persisted between page reloads for some providers (eg Google). I was unable to get Facebook to actually work. etc.

Then tonight while testing the new summary page, I started seeing this error:
> FirebaseError: [code=resource-exhausted]: Quota exceeded.
> Using maximum backoff delay to prevent overloading the backend.

This app has experienced very light traffic, so the above message is concerning.

Anyway, all of the Firebase-specific code is abstracted away behind [`DataStore`](https://github.com/bvaughn/personal-logger/blob/master/src/DataStore.js) so in theory, it should be easy enough to remove.

I [asked for recommendations about this on Twitter](https://twitter.com/brian_d_vaughn/status/934921234516664320) some time ago, and based on the responses I got, [Graph.cool](https://www.graph.cool/) seems like a good option to consider. [This article](https://blog.graph.cool/graphql-vs-firebase-496498546142) compares GraphQL to Firebase and touches on some of the pain points mentioned above (like limited querying). Another possibility is [AWS AppSync](https://aws.amazon.com/blogs/aws/introducing-amazon-appsync/).

**Comment by mbrevda** [comment]
If you want slightly better queryability, but also the real time factor, have a look at the new [Cloud Firestore](https://firebase.google.com/docs/firestore/)

**Comment by bvaughn** [comment]
I've been using Firestore, since (legacy?) Firebase was pretty horrible to use for ordered lists.

**Comment by mbrevda** [comment]
I see. (It's called the Firebase realtime database). Yeah, they are both
quite limited!

**Comment by bvaughn** [comment]
Thanks for the pointer though! I appreciate the thought.

Invested a decent amount of energy trying to get Fire* working for this project but it still feels...limited, awkward, and unintuitive to me.

Oh well. :smile:

Happy New Year buddy.

**Comment by mbrevda** [comment]
You too!

**Comment by bvaughn** [comment]
Added download/export option via #7 to support some kind of migration from Firebase to something else.

**Comment by bvaughn** [comment]
Was thinking to myself that maybe it would be nice to support a couple of backends. I'm not thrilled with Firestore, but at least it _works_ (if you overlook [hacks like this](https://github.com/bvaughn/personal-logger/blob/81d57d2c9a5e83187ea5b052887b4c2139c32b70/src/DataStore.js#L200-L205)). Maybe it would be possible to make an interface for `DataStore` and support a couple of implementations (Firebase, local storage, Graph.cool, etc.).

Maybe users could select which one to use before signing in (if we bundle-split) or maybe we could make it a build-time step.

This would be more work, and it might not be worth doing- particularly if certain backends support features that others don't. It's just a thought.

**Comment by bvaughn** [comment]
I think I've run out of patience with Graph.cool, unfortunately. The docs seems to be wrong in so many places- either outdated (referencing functions/libs that don't exist anymore) or the code they show in the tutorial fails when run in the Playground, etc...

It looks like a nice service in many ways, although the auth example is pretty heavy on the boilerplate. I just don't have the time or patience to sort through the docs at the moment, since this is a side project.

**Comment by dandv** [comment]
I also got this error while being the sole user of an app with at most 1 insert per second from the server.

> FirebaseError: [code=resource-exhausted]: Quota exceeded.
> Using maximum backoff delay to prevent overloading the backend.`

Firestore is still in beta, and [writes are capped to 1 per document per second](https://firebase.google.com/docs/firestore/quotas), so that might have triggered the error.

**Comment by joshpitzalis** [comment]
This warning just means that you guys have maxed out your free daily quota of reads and writes, you either need to bump up to a paid plan or wait a day for the quotas to be restored.

**Comment by bvaughn** [comment]
Yeah, but as the issue description mentioned- the traffic/load for this app was super minimal to have already bumped up against the free tier limit. 😊

**Comment by joshpitzalis** [comment]
It's not just traffic. There are a whole host of things that you could've maxed out if you're using apis,  cloud functions, storage, etc. If you check the usage tab on firestore/rtdb next time it happens it will take you to a dashboard that shows you exactly what maxed out.
<img width="1141" alt="app_engine_quotas_-_smart_spin_dev" src="https://user-images.githubusercontent.com/10258114/39347825-3504f986-4a12-11e8-8f96-7fbd9768becd.png">

**Comment by bvaughn** [comment]
I was using the word "traffic" very broadly to mean reads and writes. The amount of data actually stored by this app is very small.

But good suggestion, thanks for mentioning the usage tab. 😊

**Comment by ronnievsmith** [comment]
> [Google Cloud Datastore](https://cloud.google.com/datastore/pricing) offers free quota that allows you to get started at no cost. The resources and free quota amounts are listed below. If you need more quota, you must enable billing and set a spending limit for your project.
> 
> Quotas are applied daily and reset around midnight Pacific time.

![image](https://user-images.githubusercontent.com/12751611/41827084-3ef383e0-77f3-11e8-96fc-f5bfbd9cc067.png)

**Comment by thevzurd** [comment]
Is there a limit on Cloud messages as well ? I cant find any documentation on the quota of messages.
